crates/core/c8y_api/src/smartrest/message.rs

```diff
@@ -52,16 +52,69 @@ pub fn sanitize_for_smartrest(input: Vec<u8>, max_size: usize) -> String {
         .collect()
 }
 
-/// Get the last line of input.
+/// Generate a failure reason of operation with the following format.
+/// <Last line of input>
+/// (line break)
+/// <All records on input>
+///
+/// If the input has only one line, returns the line only.
 /// If the input is empty or contains invalid UTF-8, it returns an empty String.
-/// The last line is ensured to be SmartREST compatible.
-pub fn get_last_line_for_smartrest(input: Vec<u8>, max_size: usize) -> String {
+/// The output is ensured to be SmartREST compatible.
+pub fn get_failure_reason_for_smartrest(input: Vec<u8>, max_size: usize) -> String {
     let input_string = String::from_utf8(input).unwrap_or_else(|err| {
         error!("The input contains invalid UTF-8: {err}");
         String::default()
     });
     let last_line = input_string.lines().last().unwrap_or_default();
-    sanitize_for_smartrest(last_line.as_bytes().to_vec(), max_size)
+    let failure_reason = match input_string.lines().count() {
+        0 | 1 => last_line.to_string(),
+        _ => format!("{}\n\n{}", last_line, input_string.as_str()),
+    };
+    sanitize_for_smartrest(failure_reason.as_bytes().to_vec(), max_size)
+}
+
+/// Split MQTT message payload to multiple SmartREST messages.
+pub fn collect_smartrest_messages(data: &str) -> Vec<String> {
+    let mut stack: Vec<char> = Vec::new();
+    let mut smartrest_messages: Vec<String> = Vec::new();
+    let mut is_inside = false; // Inside an outermost double quote block or not.
+
+    for i in 0..data.len() {
+        let current = data.chars().nth(i).unwrap(); // Safe: i is always within the data.len()
+        if current == '"' {
+            // Check the double quote is escaped by a backslash or not.
+            // If it is escaped, it means this double quote is inside another double quote block.
+            let is_escaped = matches!(data.chars().nth(i - 1), Some('\\'));
+            if is_escaped {
```

Comment:
Since you are implementing a state machine, I would add a new flag similar to `is_inside`.

```suggestion
if current = '\\" {
   is_escaped = true;
} else if current == '"' {
   if is_escaped {
```

Doing so, you don't have to look backward neither to pop chars from the stack.

Reply:
I agree to promote `current = '\\'` to one if condition. However, we can't simply remove all backslashes.

e.g.
If a received SmartREST is
```
511,rina0005,"echo \\"hello\\""
```
then, we need to trim backslashes only before `"` like
```
511,rina0005,"echo \"hello\""
```
because the output must be
```
echo "hello"
```